cpp/src/arrow/util/async_generator.h

```diff
@@ -1195,25 +1199,48 @@ class BackgroundGenerator {
             ClearQueue();
             queue.push(spawn_status);
           }
+          task_finished.MarkFinished();
         }
       }
     }
 
     internal::Executor* io_executor;
     Iterator<T> it;
+    bool started;
     bool running;
     bool finished;
+    bool should_shutdown;
     int max_q;
     int q_restart;
     std::queue<Result<T>> queue;
     util::optional<Future<T>> waiting_future;
     util::Mutex mutex;
+    Future<> task_finished;
+  };
+
+  struct Cleanup {
+    explicit Cleanup(State* state) : state(state) {}
+    ~Cleanup() {
+      Future<> finish_fut;
+      {
+        auto lock = state->mutex.Lock();
+        if (!state->started) {
```

Comment:
I think we need only `state->running` since that also covers the case that the background task is not currently running, but has run in the past (which this check misses).

Reply:
I don't think running works.  Consider the cleanup happens while the background thread is about to call `waiting_future.MarkFinished(next);` and the queue has just filled up so `running` is false.  If we don't wait and just return then it is possible the callback accesses deleted state.

Reply:
Hmm, you're right, I got myself mixed up before. I guess the concern is actually - what if (a) the background task is currently running and (b) it's not the first time it's run - it looks like task_finished will have been completed the first time we ran the background task, so this won't actually wait for the background task to end. Maybe task_finished needs to be reinitialized when the task starts?

Reply:
`task_finished` should only be marked complete on the last execution of `Task` (e.g. `Task` should never be relaunched once `state->finished` is true).

Reply:
I think you're right though that it's not properly taking into account what happens if cleanup runs and task isn't running.